### PR TITLE
Added ability to accept Buffers as files

### PR DESCRIPTION
**Pull request by drbobbeaty**
We were really looking to use the Buffer capability of FormData, and the
only real issues we ran into were the ability to detect them, which
we've added, and the fact that the Buffers will have to have additional
properties assigned to them: .name and .contentType. If the user adds
these to the Buffer, once created with the contents of a file, then they
can simply put it in the files: [] argument to the function and it will
work just fine.

### CODE REVIEW

**Comment by erosiec** [comment]
Thanks for this! I'll get this reviewed internally and report back.

File: package.json
Comment (BHSPitMonkey) [inline package.json]: I don't think we can upstream this particular change :sweat_smile:

File: lib/HelloSignResource.js
Comment (BHSPitMonkey) [inline lib/HelloSignResource.js]: I'm commenting here as someone who doesn't regularly contribute to this repository, but I'm a little surprised this change is necessary; [`fs.createReadStream`](https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options) appears to accept both local file paths _and_ Buffers already, so I'd think that the existing code should work?

**Review by BHSPitMonkey — DISMISSED** [review]

File: package.json
Comment (drbobbeaty) [reply package.json]: Sorry... so true. 👍  I've changed it back to the published package name.

File: lib/HelloSignResource.js
Comment (drbobbeaty) [reply lib/HelloSignResource.js]: Ah... that's something that is interesting about `fs.createReadStream()` - the Buffer, or String, has to be a name of a file on the local filesystem - it just can be passed in in different ways. But it's _always_ a local file that will be read.

What we wanted was to use the output of a PDF filler (`pdf-fill-form`), which will output the contents of the PDF to a Buffer, and go _directly_ into the creation of a `SignatureRequest`. Without this change, we would have had to write the Buffer's contents to a temp file on the system, and then use that temp filename to read it in with the existing code.

We just didn't want to have that intermediate step as we already had the file's contents.

Interestingly enough, the `FormData()` library _can_ handle a Buffer for the contents of the file to put into a `POST` message. But we had to add two things: The Buffer had to also have two properties: `.name` and `.contentType`. WIth those, and this change, we got exactly what we needed. 🙂

Comment (BHSPitMonkey) [reply lib/HelloSignResource.js]: Oh, thanks for explaining! It wasn't intuitive to me that `createReadStream` still interprets Buffer contents as a *filesystem path*, but it makes sense given the module that it lives in :smile:

I'll double-check with one of our regular Node maintainers to see if we can safely land this change; Thanks again!

Comment (nathanbuchar) [inline lib/HelloSignResource.js]: Hey @drbobbeaty! Thanks for the PR :smiley:

Glancing through this, I actually think we may want to swap these two conditions:

```suggestion
                if (Buffer.isBuffer(data.files[i])) {
                    form.append("file[" + i + "]", data.files[i]);
                } else {
                    form.append("file[" + i + "]", fs.createReadStream(data.files[i]));
                }
```

In the past we never checked the type of `data.files[i]`, but this new code is specifically checking if it's a string. I think to be safe, we should instead check if it's a buffer first, and then fall back to the old code as a catch-all.

Comment (drbobbeaty) [reply lib/HelloSignResource.js]: @nathanbuchar  I made the requested change, are there any _other_ changes that are needed to unblock this?

**Comment by agelormini** [comment]
When this feature will be implemented? It will be very useful since will avoid to write files on disk before sending, which can be very good for performances especially when sending many files in sequence.

**Comment by drbobbeaty** [comment]
@coinnect-project as the initial author of the PR, this is already in our own forked copy of the NPM library:

```
    "@flexbase-eng/hellosign-sdk": "^1.7.0",
```

and that's available for you to use until the issues are worked out here.

**Comment by josiahwiebe** [comment]
This is a great PR – particularly useful if you're working in an ephemeral serverless environment and are unable to write a temp file to disk (you also wouldn't want to as you already have the contents of the file as a buffer). Looks like it's been attempted in #25 but was also never merged.

@nathanbuchar any progress on merging this in? I'm currently maintaining my own version with a similar solution but would much prefer to use the official package if I could.

**Comment by josiahwiebe** [comment]
Thought about this a bit more. 

The greater goal of allowing buffers is still 💯. I'm not sure about the implementation. The format discussed here requires reading the file and type from the Buffer, but it is somewhat inflexible. What if you could provide a file Object with those properties, along with the raw Buffer?

Something along these lines:

```js
{
  files: [
    {
      name: "NDA.pdf",
      data: <Buffer ...>,
    },
  ],
}
```

**Comment by drbobbeaty** [comment]
@josiahwiebe I'll admit that adding the attributes to `Buffer` is not "clean"... sadly, they are needed because the docs for `FormData` from the npm site say this should work:

```javascript
var form = new FormData();
form.append('my_field', 'my value');
form.append('my_buffer', new Buffer(10));
form.append('my_file', fs.createReadStream('/foo/bar.jpg'));
```

and the code in question:

```javascript
            for(var i=0; i<data.files.length;i++) {
                if (Buffer.isBuffer(data.files[i])) {
                    form.append("file[" + i + "]", data.files[i]);
                } else {
                    form.append("file[" + i + "]", fs.createReadStream(data.files[i]));
                }
            }
```

is really just doing the latter two of those examples. 🙂

So what's a library to do?

* You can certainly say "`FormData` is what it is - docs or not" and then maybe something like an optional checking for the filename in the `Buffer` is in order.
* You can say "We are working around a `FormData` bug/issue/edge case, and when they have it all figured out, then we'll be OK.

But hey... what about this? What if the suggestion allows for an "augmented" `Buffer`, and also allows for the Object model if it's not really a `Buffer`?

```javascript
            for(var i=0; i<data.files.length;i++) {
                if (Buffer.isBuffer(data.files[i])) {
                    form.append("file[" + i + "]", data.files[i]);
                } else if (typeof data.files[i] === 'object') {
                    const { name, contentType, data } = data.files[i];
                    let content = data;
                    content.name = name || 'file_upload';
                    content.contentType = contentType || 'application/pdf';
                    form.append("file[" + i + "]", data.files[i]);
                } else {
                    form.append("file[" + i + "]", fs.createReadStream(data.files[i]));
                }
            }
```

This is just written here, but it allows for folks to use Buffers - or the Object, and defaults there for the name and content type?

Just an idea... If I were really typing this into code, I'd probably put more on the Object test - like the existence of the `data`, but you get the picture. 🙂

**Comment by jtreminio-dropbox** [comment]
Please use the new OpenAPI SDK: https://www.npmjs.com/package/@dropbox/sign

**Pull request closed without merging**
